Add tablespace OID to gp_relation_node and its index.

ashwinstar opened the pull request:
Relfilenode is only unique within a tablespace. Across tablespaces same
relfilenode may be allocated within a database. Currently, gp_relation_node only
stores relfilenode and segment_num and has unique index using only those fields
without tablespace. So, it breaks for situations where same relfilenode gets
allocated to table within database.

OpenQuestions:
1] Is it better to have tablespaceOID as first column in index or second, or doesn't matter ?
2] Coded alternative patch as well to use OID instead of relfilenode in gp_relation_node. But the AlterTable rewrite case code becomes extremely messier and heavier as need to swap Oid for multiple entries in gp_relation_node. Increases bloat on gp_relation_node. With relfilendoe currently do not have to swap. Plus, clarity point of view also tablespace, relfilenode, segment file num seems more logical vs OID, segment filenum. But wish to hear broader thoughts ?

hlinnaka commented on src/backend/utils/cache/relcache.c:
```diff
@@ -492,10 +484,16 @@ FetchGpRelationNodeTuple(
 
+	/*
+	 * gp_relation_node stores tablespaceOId in pg_class fashion, hence need
+	 * to fecth the similar way.
```
typo: "fecth the ..." -> "fetch in a ..."

hlinnaka commented on src/include/catalog/gp_persistent.h:
```diff
@@ -187,6 +190,7 @@ typedef FormData_gp_relation_node *Form_gp_relation_node;
  */
 #define Schema_gp_relation_node_index \
+{ GpRelationNodeRelationId, {"tablespace_oid"}, 	26, -1,	4, 1, 0, -1, -1, true, 'p', 'i', false, false, false, true, 0 }, \
 { GpRelationNodeRelationId, {"relfilenode_oid"}, 	26, -1,	4, 1, 0, -1, -1, true, 'p', 'i', false, false, false, true, 0 }, \
```
need to adjust the attribute numbers  in the other lines here.

hlinnaka commented on src/include/catalog/gp_persistent.h:
```diff
@@ -160,6 +162,7 @@ typedef FormData_gp_relation_node *Form_gp_relation_node;
  */
 #define Schema_gp_relation_node \
+{ GpRelationNodeRelationId, {"tablespace_oid"}, 									26, -1,	4, 1, 0, -1, -1, true, 'p', 'i', true, false, false, true, 0 }, \
 { GpRelationNodeRelationId, {"relfilenode_oid"}, 									26, -1,	4, 1, 0, -1, -1, true, 'p', 'i', true, false, false, true, 0 }, \
```
need to adjust attribute numbers here

hlinnaka commented on src/backend/access/appendonly/appendonlyam.c:
```diff
@@ -547,10 +549,11 @@ SetCurrentFileSegForWrite(AppendOnlyInsertDesc aoInsertDesc)
+				&persistentTid,
+				&persistentSerialNum))
 		{
 			elog(ERROR, "Did not find gp_relation_node entry for relation name"
```
This probably needs to also print tablespace oid

ashwinstar commented:
In src/include/catalog/gp_persistent.h
<https://github.com/greenplum-db/gpdb/pull/1451#pullrequestreview-13323685>:

> > @@ -187,6 +190,7 @@ typedef FormData_gp_relation_node *Form_gp_relation_node;
>   *  pg_attribute.h]
>   */
>  #define Schema_gp_relation_node_index \
> +{ GpRelationNodeRelationId, {"tablespace_oid"}, 	26, -1,	4, 1, 0, -1, -1, true, 'p', 'i', false, false, false, true, 0 }, \
>  { GpRelationNodeRelationId, {"relfilenode_oid"}, 	26, -1,	4, 1, 0, -1, -1, true, 'p', 'i', false, false, false, true, 0 }, \
>
> need to adjust the attribute numbers in the other lines here.
>

Oppssie.. fixed the attribute numbers. Also, removed FK dependency
incorrectly coded earlier and bumped cat version.

ashwinstar commented:
Merged.
